linuxkm/lkcapi_glue.c

```diff
@@ -911,26 +1015,104 @@ static int km_AesXtsDecrypt(struct skcipher_request *req)
+            pr_err("%s: wc_AesXtsDecrypt failed: %d\n",
                    crypto_tfm_alg_driver_name(crypto_skcipher_tfm(tfm)), err);
             return -EINVAL;
         }
 
-        err = skcipher_walk_done(&walk, walk.nbytes - nbytes);
+        err = skcipher_walk_done(&walk, 0);
+
+    } else {
+        int tail = req->cryptlen % AES_BLOCK_SIZE;
+        struct skcipher_request subreq;
+
+        if (unlikely(tail > 0)) {
+            int blocks = DIV_ROUND_UP(req->cryptlen, AES_BLOCK_SIZE) - 2;
+
+            skcipher_walk_abort(&walk);
+
+            skcipher_request_set_tfm(&subreq, tfm);
+            skcipher_request_set_callback(&subreq,
+                                          skcipher_request_flags(req),
+                                          NULL, NULL);
+            skcipher_request_set_crypt(&subreq, req->src, req->dst,
+                                       blocks * AES_BLOCK_SIZE, req->iv);
+            req = &subreq;
+
+            err = skcipher_walk_virt(&walk, req, false);
+            if (!walk.nbytes)
+                return err;
+        } else {
+            tail = 0;
+        }
+
+        err = wc_AesXtsDecryptInit(ctx->aesXts, walk.iv, walk.ivsize);
 
         if (unlikely(err)) {
-            pr_err("%s: skcipher_walk_done failed: %d\n",
+            pr_err("%s: wc_AesXtsDecryptInit failed: %d\n",
                    crypto_tfm_alg_driver_name(crypto_skcipher_tfm(tfm)), err);
-            return err;
+            return -EINVAL;
         }
-    }
 
+        while ((nbytes = walk.nbytes) != 0) {
+            if (nbytes < walk.total)
+                nbytes &= ~(AES_BLOCK_SIZE - 1);
+
+            err = wc_AesXtsDecryptUpdate(ctx->aesXts, walk.dst.virt.addr,
+                                         walk.src.virt.addr, nbytes,
+                                         walk.iv);
+
+            if (unlikely(err)) {
+                pr_err("%s: wc_AesXtsDecryptUpdate failed: %d\n",
+                       crypto_tfm_alg_driver_name(crypto_skcipher_tfm(tfm)), err);
+                return -EINVAL;
+            }
+
+            err = skcipher_walk_done(&walk, walk.nbytes - nbytes);
+
+            if (unlikely(err)) {
+                pr_err("%s: skcipher_walk_done failed: %d\n",
+                       crypto_tfm_alg_driver_name(crypto_skcipher_tfm(tfm)), err);
+                return err;
+            }
+        }
+
+        if (unlikely(tail > 0 && !err)) {
+                struct scatterlist sg_src[2], sg_dst[2];
+                struct scatterlist *src, *dst;
+
+                dst = src = scatterwalk_ffwd(sg_src, req->src, req->cryptlen);
+                if (req->dst != req->src)
+                        dst = scatterwalk_ffwd(sg_dst, req->dst, req->cryptlen);
+
+                skcipher_request_set_crypt(req, src, dst, AES_BLOCK_SIZE + tail,
+                                           req->iv);
+
+                err = skcipher_walk_virt(&walk, &subreq, false);
+                if (err)
```

Comment:
unlikely(err), and indent for return seems too large

Reply:
ah good catch, the indentation needs redo there.